### PR TITLE
Run thread tagging on multiple threads

### PR DESCRIPTION
**Pull request by ashamis**

### CODE REVIEW

**Comment by cimetrics** [comment]
threading_macing_4@[6086 aka 20200318.15](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=6086&view=results) vs master ewma over 30 builds from [5800](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=5800&view=results) to [6077](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=6077&view=results)
images

**Comment by ashamis** [comment]
This was meant to be draft. There are still a couple things left to finish.

File: src/node/channels.h [hunk 1]
```diff
@@ -149,6 +153,29 @@ namespace ccf
       ctx.free_ctx();
     }
 
+    bool GetAndUpdateNonce(const GcmHdr& header, RecvNonce& nonce)
```
Comment (eddyashton) [inline src/node/channels.h]: `get_and_update_nonce`

File: src/node/nodetonode.h [hunk 2]
```diff
@@ -118,7 +118,19 @@ namespace ccf
     }
 
     template <class T>
-    CBuffer recv_authenticated_with_load(const uint8_t*& data, size_t& size)
+    bool GetRecvNonce(const uint8_t* data, size_t size, RecvNonce& nonce)
```
Comment (eddyashton) [inline src/node/nodetonode.h]: `get_recv_nonce`

File: src/consensus/pbft/pbft.h [hunk 3]
```diff
@@ -617,10 +619,16 @@ namespace pbft
           auto tmsg = std::make_unique<enclave::Tmsg<RecvAuthenticatedMsg>>(
             &recv_authenticated_msg_cb, std::move(d), this);
 
-          uint16_t tid =
-            enclave::ThreadMessaging::get_execution_thread(hdr.from_node);
+          ccf::RecvNonce recv_nonce;
+          channels->template get_recv_nonce<PbftHeader>(
+            tmsg->data.d.data(), tmsg->data.d.size(), recv_nonce);
+          if (recv_nonce.tid >= enclave::ThreadMessaging::thread_count)
```
Comment (ashamis) [inline src/consensus/pbft/pbft.h]: this should be modded

File: src/consensus/pbft/pbft.h [hunk 4]
```diff
@@ -617,10 +619,16 @@ namespace pbft
           auto tmsg = std::make_unique<enclave::Tmsg<RecvAuthenticatedMsg>>(
             &recv_authenticated_msg_cb, std::move(d), this);
 
-          uint16_t tid =
-            enclave::ThreadMessaging::get_execution_thread(hdr.from_node);
+          ccf::RecvNonce recv_nonce;
```
Comment (ashamis) [inline src/consensus/pbft/pbft.h]: make this a return value

File: src/node/channels.h [hunk 5]
```diff
@@ -162,6 +186,12 @@ namespace ccf
       key->encrypt(header.get_iv(), nullb, aad, nullptr, header.tag);
     }
 
+    void get_and_update_nonce(const GcmHdr& header, RecvNonce& nonce)
```
Comment (ashamis) [inline src/node/channels.h]: return nonce

Comment (ashamis) [reply src/node/channels.h]: remove update

**Pull request merged**